src/main/scala/com/example/TaskActor.scala

```diff
@@ -0,0 +1,245 @@
+  final case class CancelTask(id: Int, replyTo: ActorRef[TaskActorResponse]) extends TaskMessage
+  final case class TaskActorResponse(message: String)
+
+  // Internal Messages
+  final case class TaskDone(id: Int) extends TaskMessage
+  final case class TaskFailed(id: Int, exception: Throwable) extends TaskMessage
+
+  final case class TaskActorContext(runningTasks: Map[Int, UniqueKillSwitch],
+                                    taskQueue: Queue[Task],
+                                    allTasks: Map[Int, Task])
+
+
+  def extractEntityData(response: HttpResponse): Source[ByteString, _] =
+    response match {
+      case HttpResponse(OK, _, entity, _) => entity.dataBytes
+      case notOkResponse =>
+        Source.failed(new RuntimeException(s"illegal response $notOkResponse"))
+    }
+
+
+  def toJson(map: Map[String, String])(implicit jsWriter: JsonWriter[Map[String, String]]): JsValue = jsWriter.write(map)
+
+
+  def apply(): Behavior[TaskMessage] = Behaviors.setup(ctx => {
+    ctx.log.info("Starting")
+    taskActorBehavior(ctx, TaskActorContext(Map.empty, Queue.empty, Map.empty))
+  })
+
+  private def taskActorBehavior(ctx: ActorContext[TaskMessage], taskCtx: TaskActorContext): Behavior[TaskMessage] = {
+    implicit val mat: Materializer = Materializer(ctx)
+
+    val MAX_CAPACITY = ctx.system.settings.config.getInt("downloader.max-capacity")
+    val PARALELLISM = ctx.system.settings.config.getInt("downloader.parallelism")
+    val OUTPUT_FOLDER = ctx.system.settings.config.getString("downloader.output-folder")
+    val DELAY_SECONDS = ctx.system.settings.config.getDuration("downloader.delay")
+
+    //val KAGGLE_USERNAME = ctx.system.settings.config.getString("kaggle.username")
+    //val KAGGLE_API_KEY = ctx.system.settings.config.getString("kaggle.apikey")
+
+    val downloadFlow: Flow[Task, IOResult, NotUsed] = Flow[Task].mapAsyncUnordered(PARALELLISM) { task =>
+      val request = Get(task.csvUri)
+      //.withHeaders(Authorization(BasicHttpCredentials(username = KAGGLE_USERNAME, password = KAGGLE_API_KEY)))
+      val source = Source.single(request)
+      val filePath = Paths.get(s"$OUTPUT_FOLDER/${task.id}.json")
+
+      import spray.json.DefaultJsonProtocol._
+
+      import scala.concurrent.duration._
+
+      source
+        .delay(DELAY_SECONDS.toSeconds seconds)
+        .mapAsyncUnordered(1)(Http()(ctx.system).singleRequest(_))
+        .flatMapConcat(extractEntityData)
+        .via(CsvParsing.lineScanner())
+        .via(CsvToMap.toMapAsStrings())
+        .map(toJson)
+        .map(_.compactPrint)
+        .intersperse("[", ",\n", "]")
+        .map(ByteString(_))
+        .withAttributes(ActorAttributes.dispatcher("csv-download-dispatcher"))
+        .runWith(FileIO.toPath(filePath))
+    }
+
+    def runTask(task: Task): UniqueKillSwitch = {
+      Source(List(task))
+        .via(downloadFlow)
+        .viaMat(KillSwitches.single)(Keep.both)
```

Comment:
Such usage of KillSwitch won't stop `downloadFlow` running inside Future/mapAsync

Reply:
Thanks for letting me know! Can you elaborate how could this be an issue and how to solve it, I couldn't find any more info apart from this
https://doc.akka.io/docs/akka/current/stream/stream-dynamic.html#controlling-stream-completion-with-killswitch

Reply:
The point of task cancelation is to not waste resources when not needed. In this case you have basically two flows running in parallel. When you use kill switch on the outer level, you only stop the "waiting for results" flow, while the `downloadFlow` which you run via `run` is executed as `Future`.

One solution which comes to my mind is to merge those two flows together, that way single KillSwitch inserted just after `flatMapConcat` would stop consuming new bytes from the HTTP.

I would also try to somehow get rid of `mapAsync` and `run` in the flow blueprint. If not possible, you could use SharedKillSwitch, which could be used to stop both parallel Flows.